File: 백준/레벨Gold III/10986. 나머지 합/나머지 합.py
from sys import stdin

input = stdin.readline

def range_sum():
    # 누적 합 초기화
    N, M = map(int, input().split())
    nums = list(map(int, input().split()))

    pre = [0] * (N + 1)
    for i in range(N):
        pre[i + 1] = (pre[i] + nums[i]) % M
    
    # 나머지 배열 초기화
    r = [0] * M
    for i in range(1, N+1):
        r[pre[i]] += 1
    
    # 조합
    count = r[0] # 누적합도 구간합
    for i in range(M):
        count += (r[i] * (r[i]-1)) // 2 # nC2 = n * n-1 / 2

    print(count)

range_sum()

File: 백준/레벨Gold III/10986. 나머지 합/나머지 합.cc
#include <bits/stdc++.h>

using namespace std;

void rangeSum() {
    // 누적 합 초기화
    int N, M;
    cin >> N >> M;

    vector<int> nums(N);
    for (int i = 0; i < N; i++) {
        cin >> nums[i];
    }

    vector<int> pre(N + 1, 0);
    for (int i = 0; i < N; i++) {
        pre[i + 1] = (pre[i] + nums[i]) % M;
    }

    // 나머지 배열 초기화
    vector<int> r(M, 0);
    for (int i = 1; i <= N; i++) {
        r[pre[i]]++;
    }

    // 조합
    long long count = r[0]; // 누적합도 구간합
    for (int i = 0; i < M; i++) {
        count += (long long) r[i] * (r[i] - 1) / 2; // nC2 = n * (n-1) / 2
    }

    cout << count;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    rangeSum();
}